### DISCUSSION TITLE
try var

### DISCUSSION THREAD

**Discussion by joycestick** [dotnet/csharplang#4683] [Language Ideas]
I find I often have a `try` block with just a function call in it, the return value of which I need outside the scope of the `try`, e.g.

```cs
double x;
try
{
    x = GetValue();
}
catch (SomeException e)
{
    // deal with exception
}
// do something with x
```

The only real drawbacks of the above code are explicit type declaration (I'd rather use `var`) and number of lines of code, so this suggestion is only a bit of sugar, but I wondered if something like the following would be possible:

```
try var x = GetValue();
catch (SomeException e)
{
    // deal with exception
}
// do something with x
```

**EDIT**: It has been noted that in both cases `x` will be uninitialised at the point of `// do something with x` if `GetValue()` throws an exception and the `catch` block doesn't assign `x`. This suggestion does not attempt to address that issue.

I've had a good look for this suggestion as already existing and can't find anything, but apologies if it's a duplicate.

Also, any suggestion around try-catch seems to be pretty contentious, but I'm hoping this one isn't as offensive as asking for an `On Error Resume Next` equivalent.

**Comment by canton7** [comment]
I think [expression statements](https://github.com/dotnet/csharplang/issues/3086) might cover this case: it's mentioned as a [case here](https://github.com/dotnet/csharplang/issues/3086#issuecomment-576462742).

If I'm reading it right, something like:

```cs
double x = {
    try
    {
        break GetValue();
    }
    catch (SomeException e)
    {
        // ...
    }
}
```

... although the syntax seems to be somewhat up for grabs.

**Reply by joycestick** [reply]
But isn't that more lines of code and still no type inference? There may be a way to use expression statements - I've not fully absorbed how they could be used effectively yet.

**Reply by canton7** [reply]
I imagine you could write `var x = ...` if you wanted to: the compiler's got all of the information it needs.

See how languages such as Ruby use them:

```ruby
x = begin
  get_value
rescue e
  # ...
end
```

**Comment by bernd5** [comment]
Or [improved lambdas](https://github.com/dotnet/csharplang/blob/main/proposals/lambda-improvements.md):

Today you can write:
```csharp
using System;

double x = new Func<double>(() =>{
    try
    {
        return 42;
    }
    catch (Exception e)
    {
        // ...
        throw;
    }
})();
```

But in the future (maybe csharp 10)
It could be simply:

```csharp
using System;

double x = () =>{
    try
    {
        return 42;
    }
    catch (Exception e)
    {
        // ...
        throw;
    }
}();
```

**Reply by Richiban** [reply]
IFFEs are an awfully expensive way of doing this kind of thing though

**Reply by bernd5** [reply]
A "self calling lambda" needs never any heap allocation. It could be perfectly inlined.
But currently roslyn does implement this like all other lambdas with allocation overhead. That could be changed.
And we were talking about the syntax - not the actual implementation.

**Reply by bernd5** [reply]
See how efficiently it could be implemented: [c++ implementation](https://godbolt.org/z/4eKcKePfY)

It is actually compiled to:

```asm
mov     eax, 42
ret
```

**Comment by spydacarnage** [comment]
The issue with all the suggestions in this thread so far is that there is no emphasis on what happens to `x` when an exception is caused.
For instance:
```cs
try var x = GetValue();
catch (SomeException e)
{
    // deal with exception
}
// x is not defined at this point
```

**Reply by canton7** [reply]
Presumably you'd get the same definite assignment rules / errors as with:

```cs
double x;
try
{
    x = GetValue();
}
catch (SomeException e)
{
    // deal with exception
}
// Error if the 'catch' block doesn't assign 'x', or rethrow
```

**Reply by jnm2** [reply]
Would be good to specify this up front.

**Reply by bernd5** [reply]
With improved lambdas YOU have to specifiy the value - without returning a value it is a compile error.

**Reply by joycestick** [reply]
Isn't that the same situation as the initial code where an exception means `x` may not be initialised? So I think you're not losing anything against the current approach.

EDIT - Sorry, delayed post. This is the same as what @canton7 is saying.

**Reply by joycestick** [reply]
Thanks, @jnm2. This would definitely be something to point out if this discussion gets enough support to put this in as a proper proposal. Worth me editing this discussion as well do you think?

**Reply by spydacarnage** [reply]
The original, however, could be
```cs
double x = 0;
try ...
```
Whereas you lose that with:
```cs
try double x = ...
```

**Reply by smoothdeveloper** [reply]
I concur with @spydacarnage concern, it should give indication of using unassigned local in this case, using CS0165 error.

```csharp
try var x = GetValue();
catch (SomeException e)
{
    // deal with exception
}
// if x is not assigned in the catch, or the catch is not throwing or returning:
// CS0165: use of unassigned local variable 'x'
```

So maybe:

```csharp
var x =
  try => GetValue()
  catch (e) => ExceptionHandler(e).Then().GetValueWhenItFailed()
  // also need to think about finally...
;
```

This can't be an error since the catch is forced to return an expression of same type of the try branch (or it could be either throwing or returning altogether).

It is nice idea to have an alternative to try/catch being only a statement block and turning it into an expression.

There are other changes that made it to C#, like the switch expression or expression bodied members, so finding an approach which work with try/catch: :+1:.

**Reply by jnm2** [reply]
@joycestick Only if you think other people will notice the same thing and miss this comment thread (or not even read down this far).

**Reply by joycestick** [reply]
@spydacarnage Yeah agreed. The situation where you want to preset a value before executing the `try` isn't something I'd originally intended on replacing. I'd just use the currently available syntax in that case.

Of course if you wanted to ensure `x` was initialised you could always make the proposed syntax of:

```
try var x = GetValue();
```

equivalent to:

```cs
InferredType x = default;
try
{
    x = GetValue();
}
```

but I think setting a variable to its default in this way should be explicitly the job of the developer in the `catch` block if that's what they want to happen.

**Comment by djthegr8** [comment]
Another possibility is, and this solely applies to the case where the exception caught is irrelevant, we could set x to null if an exception is thrown. A major throwback of this would be that a returned `null` would be the same as a thrown exception, plus you can't set a non-nullable type in this... 
```cs
// GetValue does not throw an exception, x would be equal to whatever it should be
try var x = GetValue();
// ThrowValue throws an exception, and y is set to null
try var y = ThrowValue();
```

**Reply by HaloFour** [reply]
> and this solely applies to the case where the exception caught is irrelevant

Which should be incredibly rare and not a behavior to be encouraged by the language.

**Reply by djthegr8** [reply]
I agree that it may not be encouraged, and *is* bad code in my opinion, but I disagree with the rarity, it is pretty often that you see blocks of code like 
```cs
try {
// Do something like assign a variable
var x = GetValue();
return x % 2 == 0 ? "even" : "odd";
} catch {
// Let it be
} 
```

**Reply by HaloFour** [reply]
I'd suggest that if you're seeing that more than 1% of the time that something is glaringly wrong.  Analyzers will typically flag such use of `try`/`catch` as inappropriate.  If the language were to make it easier to write it would be tantamount to encouraging use of that pattern.  In my opinion the verbosity (as little as it is) should remain if only to give the developer pause to consider whether that kind of exception handling is appropriate.

**Reply by djthegr8** [reply]
Makes sense... VS doesn't really mind this, and R# ensures that you write a comment as to why you're ignoring the Exception, so that's that...

**Reply by theunrepentantgeek** [reply]
> it is pretty often that you see blocks of code like

Your experience differs widely from mine - I can't remember the last time I reviewed or wrote code that simply ignored an exception like that. 

In fact, I'm pretty sure that any PR containing that would be promptly rejected with a comment instructing me to fix it.

I think @HaloFour sumarized it well - this is not a coding practice that should be encouraged, so the language shouldn't be making it easier to do.

**Reply by differentrain** [reply]
> Your experience differs widely from mine - I can't remember the last time I reviewed or wrote code that simply ignored an exception like that.
 
In my experience, I have to write these code:
In library project, all exceptions will be thrown and handled, but in some UI project, these exceptions are "converted" to `null`.
So I like these code:
```
// GetValue does not throw an exception, x would be equal to whatever it should be
try var x = GetValue();
// ThrowValue throws an exception, and y is set to default value
try default y = ThrowValue();
// ThrowValue throws an exception, and this exception is thrown
try var z = ThrowValue();
catch (exception e)
{
  //todo
}
```

**Reply by theunrepentantgeek** [reply]
With due apologies for being slightly off topic ...

> in some UI project, these exceptions are "converted" to null.

This reads as though you have applications that blithely discard any invalid user input without giving the user any opportunity to correct things. That sounds actively user hostile to me - and is an approach the language absolutely positively should not endorse by enshrining with specific syntax.

If you absolutely must have such a thing, you can easily achieve it with a simple helper method that you bring into scope with **using status** and use where required:

``` csharp
public static T Try<T>(Func<T> func, T fallback) 
{
    try
    {
        return func();
    }
    catch
    {
        return fallback;
    }
}
```

**Comment by gulshan** [comment]
Some prior discussion here- #2734
